daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -166,21 +168,48 @@ private[engine] final class Preprocessor(
     }
 
   private[engine] def preprocessApiCommand(
-      pkgMap: Map[Ref.PackageName, Ref.PackageId],
+      pkgResolution: Map[Ref.PackageName, Ref.PackageId],
       cmd: command.ApiCommand,
   ): Result[speedy.Command] =
-    safelyRun(pullPackage(pkgMap, List(cmd.typeRef))) {
-      commandPreprocessor.unsafePreprocessApiCommand(pkgMap, cmd)
+    safelyRun(pullPackage(pkgResolution, List(cmd.typeRef))) {
+      commandPreprocessor.unsafePreprocessApiCommand(pkgResolution, cmd)
     }
 
+  def buildPackageResolution(
+      packageMap: Map[Ref.PackageId, (Ref.PackageName, Ref.PackageVersion)] = Map.empty,
+      packagePreference: Set[Ref.PackageId] = Set.empty,
+  ): Result[Map[PackageName, PackageId]] =
+    packagePreference.foldLeft(Result.done(Map.empty[Ref.PackageName, Ref.PackageId]))(
+      (acc, pkgId) =>
+        acc.flatMap(m =>
+          packageMap.get(pkgId) match {
+            case Some((pkgName, _)) =>
+              m.get(pkgName) match {
+                case None =>
+                  Result.done(m.updated(pkgName, pkgId))
+                case Some(pkgId0) =>
+                  ResultError(
+                    Error.Preprocessing.Internal(
+                      NameOf.qualifiedNameOfCurrentFunc,
+                      s"package $pkgId0 and $pkgId have the same name",
```

Comment:
I think it would be helpful for us to show the triplet (PackageId, PackageName, PackageVersion) for the conflicting packages

Reply:
This is an internal error. So I do not think we need more info. 
Tudor is responsible to enforce that.

Reply:
Indeed, the participant will pre-validate these